### ISSUE TITLE
Memory leak regression in DefaultDataType cache

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11345]
From the user group:
https://groups.google.com/g/jooq-user/c/4tM-YEuqISw

> We want to report an issue with the latest release. We upgraded from 3.14.4 to 3.14.6 and then experienced a memory leak where the heap was slowly filling up with instances of org.jooq.impl.DefaultDataType instances.
> 
> We took a java heap dump of our service (a spring-boot service) and analysed it with Eclipse Memory Analyzer Tool. It showed that 70% of the memory was accumulated in a java.util.LinkedHashMap which contained these DefaultDataType objects. We didn't dig into this further, but the problem completely disappeared once we downgraded back to 3.14.4 so our conclusion is that the issue was with the new release.

**Comment by lukaseder** [comment]
A few `DataType` related improvements backported to 3.14.5 and 3.14.6 include:

- #10809 Dialect specific data type is reported from Meta, instead of SQLDataType
- #11309 ArrayDataType.toString() should indicate that it's an array type
- #5310 DataType.toString() should generate DDL type declaration fragment
- #11168 DataType.convert(Object...) and convert(Collection<?>) don't work for converted data types
- #11086 Wrong SQL generated when DomainDataType is loaded before base SQLDataType for SQLDialect.DEFAULT

**Comment by lukaseder** [comment]
The most likely candidate is https://github.com/jOOQ/jOOQ/issues/11086 with this change:

https://github.com/jOOQ/jOOQ/commit/995cce0d6e60df9f7ba53962eec4eca5d1573d47 (main branch), https://github.com/jOOQ/jOOQ/commit/402a0299dbd0780944153c2a8950060b5d4c4f79 (3.14 branch)

Backported to 3.14.5 with https://github.com/jOOQ/jOOQ/issues/11087

```patch
From 402a0299dbd0780944153c2a8950060b5d4c4f79 Mon Sep 17 00:00:00 2001
From: Lukas Eder <lukas.eder@gmail.com>
Date: Thu, 3 Dec 2020 17:18:59 +0100
Subject: [PATCH] [jOOQ/jOOQ#11086] Wrong SQL generated when DomainDataType is
 loaded before base SQLDataType for SQLDialect.DEFAULT

---
 jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java b/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
index e466576ae5..8227e14fbb 100644
--- a/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
+++ b/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
@@ -289,8 +289,8 @@ public DefaultDataType(SQLDialect dialect, Class<T> type, String typeName, Strin
 
         this.dialect = dialect;
 
-        // [#858] SQLDataTypes should reference themselves for more convenience
-        this.sqlDataType = (dialect == null) ? this : sqlDataType;
+        // [#858] [#11086] SQLDataTypes should reference themselves for more convenience
+        this.sqlDataType = (dialect == null && sqlDataType == null) ? this : sqlDataType;
         this.uType = type;
         this.typeName = TYPE_NAME_PATTERN.matcher(typeName).replaceAll("").trim();
         this.castTypeName = castTypeName == null ? this.typeName : castTypeName;
@@ -325,8 +325,8 @@ public DefaultDataType(SQLDialect dialect, Class<T> type, String typeName, Strin
         if (TYPES_BY_TYPE[ordinal].get(type) == null)
             TYPES_BY_TYPE[ordinal].put(type, this);
 
-        if (TYPES_BY_SQL_DATATYPE[ordinal].get(sqlDataType) == null)
-            TYPES_BY_SQL_DATATYPE[ordinal].put(sqlDataType, this);
+        if (TYPES_BY_SQL_DATATYPE[ordinal].get(this.sqlDataType) == null)
+            TYPES_BY_SQL_DATATYPE[ordinal].put(this.sqlDataType, this);
 
         // Global data types
         if (dialect == null)
```

**Comment by lukaseder** [comment]
It should be reproducible when using manual calls to `DataType::asConvertedDataType`, or otherwise creating new data types, i.e. when using dynamic schemas, not so much with ordinary generated code.

**Comment by lukaseder** [comment]
This program seems to reproduce it:

```java
import org.jooq.impl.SQLDataType;

public class Leak {

    public static void main(String[] args) {
        for (;;) {
            SQLDataType.INTEGER.asConvertedDataType(Converters.identity(Integer.class));
        }
    }
}
```

This patch, partially reverting the previous change, against 3.14 seems to fix it:

```patch
diff --git a/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java b/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
index 729a9ec..e2ebebf 100644
--- a/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
+++ b/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
@@ -334,8 +334,8 @@
         if (TYPES_BY_TYPE[ordinal].get(type) == null)
             TYPES_BY_TYPE[ordinal].put(type, this);
 
-        if (TYPES_BY_SQL_DATATYPE[ordinal].get(this.sqlDataType) == null)
-            TYPES_BY_SQL_DATATYPE[ordinal].put(this.sqlDataType, this);
+        if (TYPES_BY_SQL_DATATYPE[ordinal].get(sqlDataType) == null)
+            TYPES_BY_SQL_DATATYPE[ordinal].put(sqlDataType, this);
 
         // Global data types
         if (dialect == null)
```

Patch against 3.15:

```patch
diff --git a/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java b/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
index 039e53a..8137f7e 100644
--- a/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
+++ b/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
@@ -328,7 +328,7 @@
             TYPES_BY_NAME[ordinal].putIfAbsent(DefaultDataType.normalise(typeName), this);
 
         TYPES_BY_TYPE[ordinal].putIfAbsent(type, this);
-        TYPES_BY_SQL_DATATYPE[ordinal].putIfAbsent(this.sqlDataType, this);
+        TYPES_BY_SQL_DATATYPE[ordinal].putIfAbsent(sqlDataType, this);
 
         // Global data types
         if (dialect == null)
```

**Comment by lukaseder** [comment]
The patch reverts a change from #11086, which may by itself introduce a new regression on that fix. The regression may be the lesser evil compared to this significant memory leak. To be verified

**Comment by lukaseder** [comment]
A slightly better patch would be to add an additional `null` check. 3.15 version:

```patch
diff --git a/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java b/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
index 039e53a..c667d0f 100644
--- a/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
+++ b/jOOQ/src/main/java/org/jooq/impl/DefaultDataType.java
@@ -328,7 +328,8 @@
             TYPES_BY_NAME[ordinal].putIfAbsent(DefaultDataType.normalise(typeName), this);
 
         TYPES_BY_TYPE[ordinal].putIfAbsent(type, this);
-        TYPES_BY_SQL_DATATYPE[ordinal].putIfAbsent(this.sqlDataType, this);
+        if (sqlDataType != null)
+            TYPES_BY_SQL_DATATYPE[ordinal].putIfAbsent(sqlDataType, this);
 
         // Global data types
         if (dialect == null)
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.7 (#11353)
